libs/@local/harpc/net/src/session/server/connection/mod.rs

```diff
@@ -0,0 +1,342 @@
+        // channel, meaning that `ConnectionDelegateTask` will stop itself.
+        // ^ this is true in theory, but we only drop `tx` if task has been finished, which is a
+        // problem, this has the potential of creating a task that just never stops. Which we do NOT
+        // want.
+        // We solve this by dropping the `tx` once the stream has finished, because we know that we
+        // won't be able to create any more transactions.
+        let (tx, rx) = mpsc::channel(self.config.per_connection_response_buffer_size.get());
+        let mut handle = tasks
+            .spawn(ConnectionDelegateTask { rx, sink }.run(cancel.clone()))
+            .fuse();
+
+        let mut tx = Some(tx);
+
+        loop {
+            select! {
+                // we use `StreamNotifyClose` here (and the double `Option<Option<T>>`), so that we don't add too many permits at once
+                // `StreamNotifyClose` is guaranteed to end once the stream is closed, and we won't poll again.
+                Some(request) = stream.next() => {
+                    match request {
+                        None => {
+                            // stream has finished
+                            finished.add_permits(1);
+
+                            // shutdown the senders, as they can't be used anymore, this indicates to the delegate tasks that they should stop
+                            // we don't cancel them, because they might still respond to the requests,
+                            // they'll cancel themselves once the handle has finished
+                            self.transactions.shutdown_senders();
+
+                            // drop the sender (as we no longer can use it anyway)
+                            // this will also stop the delegate task, once all transactions have finished
+                            tx.take();
+                        }
+                        Some(Ok(request)) => {
+                            let tx = tx.clone().expect("infallible; sender is only unavailble once the stream is exhausted");
+
+                            self.handle_request(tx, &tasks, request).await;
+                        },
+                        Some(Err(error)) => {
+                            tracing::info!(?error, "malformed request");
+                        }
+                    }
+                },
+                result = &mut handle => {
+                    match result {
+                        Ok(Ok(())) => {},
+                        Ok(Err(error)) => {
+                            tracing::warn!(?error, "connection prematurely closed");
+                        },
+                        Err(error) => {
+                            tracing::warn!(?error, "unable to join connection delegate task");
+                        }
+                    }
+
+                    finished.add_permits(1);
+                },
+                () = notify.notified() => {
```

Comment:
Similar to above: The comment below helps a lot but it would be good if `notify` would be renamed. Maybe `item_removed_notify`?

Reply:
renamed in [4adf694](https://github.com/hashintel/hash/pull/4459/commits/4adf694885abb40ac80f7f817aa2f1ae0776b701)